packages/react/src/types/templates.ts

```diff
@@ -46,37 +54,50 @@ export interface OracleInitializationBundle {
 export type OracleInitializationBundleGetter =
     () => Promise<OracleInitializationBundle>;
 
-export type OracleChangeCallback<T> = (
-    internalState: Partial<T>,
+export type OracleChangeCallback<S extends Serializable = Serializable> = (
+    internalState: S,
     initializationBundleGetter?: OracleInitializationBundleGetter,
 ) => void;
 
-export interface AdditionalRemoteOracleCreationFormProps<T> {
-    state: Partial<T>;
+export interface AdditionalRemoteOracleCreationFormProps<
+    S extends Serializable = Serializable,
+> {
+    state: S;
     kpiToken?: Partial<KPIToken>;
-    onChange: OracleChangeCallback<T>;
+    onChange: OracleChangeCallback<S>;
     navigate: NavigateFunction;
     onTx: <T extends TxType>(tx: Tx<T>) => void;
 }
 
-export type OracleRemoteCreationFormProps<T> =
-    BaseRemoteTemplateComponentProps &
-        AdditionalRemoteOracleCreationFormProps<T>;
-
-export type OracleCreationFormProps<S> = TemplateComponentProps &
+export type OracleRemoteCreationFormProps<
+    S extends Serializable = Serializable,
+> = BaseRemoteTemplateComponentProps &
     AdditionalRemoteOracleCreationFormProps<S>;
 
-export interface AdditionalRemoteKPITokenCreationFormProps {
+export type KPITokenChangeCallback<S extends Serializable = Serializable> = (
+    state: S,
+) => void;
+
+export type OracleCreationFormProps<S extends Serializable = Serializable> =
+    TemplateComponentProps & AdditionalRemoteOracleCreationFormProps<S>;
+
+export interface AdditionalRemoteKPITokenCreationFormProps<
+    S extends Serializable = Serializable,
+> {
+    state: S;
+    onChange: KPITokenChangeCallback<S>;
```

Comment:
Should we rename this to be more specific? To what `onChange` event this function is linked to in the `KPITokenCreationForm`?

Reply:
The internal state's change. I think it's probably clear enough when looking at the `KPITokenChangeCallback` type. It's the same name and functionality in the oracle template type too I think.